src/foam/parse/PropertyParser.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.parse;
+
+import java.util.List;
+
+import foam.core.PropertyInfo;
+import foam.lib.json.KeyParser;
+import foam.lib.json.Whitespace;
+import foam.lib.parse.Alt;
+import foam.lib.parse.Literal;
+import foam.lib.parse.PStream;
+import foam.lib.parse.ParserContext;
+import foam.lib.parse.ProxyParser;
+import foam.lib.parse.Seq3;
+import foam.mlang.MLang;
+import foam.mlang.predicate.Predicate;
+
+public class PropertyParser extends ProxyParser {
+  private PropertyInfo property;
+
+  public PropertyParser(PropertyInfo p) {
+    super(new Seq3(4, 3,

```

Comment:
In my latest example, you can see how we actually construct a unique expression parser for each property.  This means we don't need a generic PropertyParser like you have here.  This parser is rather large and complicated because it has to handle every possible type of property and value combination.  In my example, we don't have any PropertyParser anymore because for any given property there are only a few valid expression for it.  Numeric properties can only be less-than, greather-than, less-than-or-equal, greater-than-or-equal, or equal to a value.  String properties have different possibilties, date properties have their own set of possibilities.

By constructing a different expression parser for each property type, we can avoid a mega parser like this and having more smaller parsers that are easier to understand and maintain.  It also means the system is more extensible, as new properties types don't require updating this one class.